src/control/cmd/daos_agent/mgmt_rpc.go

```diff
@@ -216,81 +211,96 @@ func (mod *mgmtModule) getAttachInfoResp(ctx context.Context, numaNode int, sys
+		resp, err := mod.selectSecondaryAttachInfo(srvResp, mod.providerIdx)
+		if err != nil {
+			return nil, err
+		}
+
+		if len(reqProviders) != 0 && !reqProviders.Has(resp.ClientNetHint.Provider) {
+			mod.log.Errorf("requested fabric interface %q (domain: %q) does not report support for configured provider %q (idx %d)",
+				iface, domain, resp.ClientNetHint.Provider, mod.providerIdx)
+		}
+
+		return resp, nil
+	}
+
+	if len(reqProviders) == 0 || reqProviders.Has(srvResp.ClientNetHint.Provider) {
+		return srvResp, nil
+	}
+
+	mod.log.Debugf("primary provider is not supported by requested interface %q domain %q (supports: %s)", iface, domain, strings.Join(reqProviders.ToSlice(), ", "))
+
+	// We can try to be smart about choosing a provider if the client requested a specific interface
+	for _, hint := range srvResp.SecondaryClientNetHints {
+		if reqProviders.Has(hint.Provider) {
+			mod.log.Debugf("found secondary provider supported by requested interface: %q (idx %d)", hint.Provider, hint.ProviderIdx)
+			return mod.selectSecondaryAttachInfo(srvResp, uint(hint.ProviderIdx))
+		}
+	}
+
+	mod.log.Errorf("no supported provider for requested interface %q domain %q, using primary by default")
+	return srvResp, nil
+}
+
+func (mod *mgmtModule) getIfaceProviders(ctx context.Context, iface, domain string) common.StringSet {
+	providers := common.NewStringSet()
 	if iface == "" {
-		return nil
+		return providers
 	}
 
 	if domain == "" {
 		domain = iface
 	}
 
-	fis, err := mod.fabricInfo.localNUMAFabric.FindDevice(&FabricIfaceParams{
+	if fis, err := mod.getFabricInterface(ctx, &FabricIfaceParams{
 		Interface: iface,
 		Domain:    domain,
-	})
-	if err != nil {
-		mod.log.Errorf("client-requested fabric interface/domain not detected: %s", err.Error())
-		mod.log.Error("communications on this interface may fail")
-		return nil
+	}); err != nil {
+		mod.log.Errorf("requested fabric interface %q (domain %q) may not function as desired: %s", iface, domain, err)
+	} else {
+		providers.Add(fis.Providers()...)
 	}
 
-	providers := common.NewStringSet()
-	for _, fi := range fis {
-		providers.AddUnique(fi.Providers()...)
-	}
 	return providers
 }
 
-func (mod *mgmtModule) selectAttachInfo(srvResp *mgmtpb.GetAttachInfoResp, reqProviders common.StringSet) (*mgmtpb.GetAttachInfoResp, error) {
-	providers := reqProviders
-	if mod.provider != "" {
-		if len(reqProviders) > 0 && !reqProviders.Has(mod.provider) {
-			mod.log.Errorf("configured provider %q not included in requested interface's detected providers: %s", reqProviders)
-			mod.log.Error("communications on this interface may fail")
-		}
-		providers = common.NewStringSet(mod.provider)
-
-		mod.log.Debugf("using configured provider: %s", mod.provider)
+func (mod *mgmtModule) selectSecondaryAttachInfo(srvResp *mgmtpb.GetAttachInfoResp, provIdx uint) (*mgmtpb.GetAttachInfoResp, error) {
+	maxIdx := len(srvResp.SecondaryClientNetHints)
+	if int(provIdx) > maxIdx {
```

Comment:
Probably should guard against `provIdx == 0` just in case, right?

Reply:
In theory it shouldn't happen if we've gotten to this point, but if the server gave us bad data it could be possible. I'll add a check.